### ISSUE TITLE
powercap-rapl: energy_uj file to be restricted to root access

### ISSUE THREAD

**Issue by connorimes** [powercap/powercap#7]
Linux kernel development indicates that the `energy_uj` file will be restricted to root-only permissions [1] in an upcoming kernel release---probably 5.10.

The `powercap-rapl` interface currently only skips over files gracefully if they don't exist, but an open failure due to permissions will cause `powercap_rapl_init` to fail.  We need to decide if, and if so, how, to address this file mode change.

The "default" approach is to not make any changes, which might be fine for most use cases.  The `powercap-rapl` interface was primarily intended for software that needs to either read energy or change power caps with regularity, the latter of which already requires root access.  Is it common to want to access other files that don't require root privileges with such regularity that using the `powercap-rapl` interface is desirable, as opposed to just using `powercap-sysfs`?

Because the kernel developers don't seem consider this an interface change (which IMO it *is*), any solution we implement should be general for all files in a zone (including constraint files), i.e., not limited to just working around `energy_uj`, in case other such kernel changes are made in the future.

[1] https://github.com/torvalds/linux/commit/949dd0104c496fa7c14991a23c03c62e44637e71

**Comment by connorimes** [comment]
This change is released linux v5.10

**Comment by amkozlov** [comment]
omg, seriously? Intel had one year time to fix the problem [1]. and didn't find anything better then to axe power monitoring capabilities for all unprivileged programs. What's next, `sudo top`? :-/

[1] https://platypusattack.com/

**Comment by connorimes** [comment]
Indeed... Are you finding that you need a graceful degradation solution as a result of this permissions change?  My current plan is to not do anything about this.

Additionally I want to deprecate the `powercap-rapl` interface and replace it with something more general since `powercap-rapl` is not extensible, e.g., I can't cleanly add the new PL4 constraint without breaking the API.  The challenge is that a more general solution seems to me to require implementing a tree structure.  First: there are different approaches to implementing trees in C, and all are somewhat unwieldy for users.  Second: implementing a tree structure will almost certainly require the API to take on memory management responsibility, which I have hoped to avoid.  Finally: navigating a tree structure is less efficient than the constant time accesses `powercap-rapl` provides, although that constant-time access came with assumptions about the zone and constraint structures, which is exactly what makes the current API problematic.  I thus far haven't had a pressing need for a new API, so this task has been sitting on the backburner for months (even years now), but I don't want to sit on it forever.  /Sigh...

**Comment by amkozlov** [comment]
> Indeed... Are you finding that you need a graceful degradation solution as a result of this permissions change? My current plan is to not do anything about this.

Well so far, I just check if `energy_uj` file is readable and disable power monitoring if it's not. What else can I do from userspace? 
I also don't want to ask users to run my program under root or to revert RAPL file permissions (even if they can). I think the ideal solution would be that some privileged code (kernel module? daemon?) exports "safe" aggregated energy data under `/proc/energy` or similar. And then (optionally) more detailed/fine-grained measurements under `/proc/<pid>/`, such that I can safely read data for my own processes. Energy can be attributed to processes [according to their CPU usage](https://github.com/Breakend/experiment-impact-tracker/blob/master/experiment_impact_tracker/cpu/intel.py#L126). This shouldn't be too hard to implement, although I'm not an expert in Linux kernel programming...

Some good news: AMD support for powercap [is underway](https://www.phoronix.com/scan.php?page=news_item&px=AMD-Zen-PowerCap-RAPL-5.11) and energy monitoring is [already available](https://www.kernel.org/doc/html/latest/hwmon/amd_energy.html) via `amd_energy` driver, even with integrated polling!

**Comment by connorimes** [comment]
There isn't much you can do except ask for one of those two things.  From the perspective of the powercap library in this repo, users shouldn't have to interact with the filesystem structure at all - they shouldn't need to check file permissions in advance.  The library initialization either succeeds or fails for some reason given in `errno`.

What you're describing is essentially the powercap sysfs file interface - a "safe" way (via a kernel module like `intel_rapl`, `intel_rapl_mmio`, and now apparently `amd_energy`) to get power readings without users having to parse MSR bits directly (which linux is also restricting more and more, even for `root`).  Whether the readings come from the current powercap interface or something new, the same fundamental security problem likely still exists, so linux would lock down the permissions.

**Comment by amkozlov** [comment]
> What you're describing is essentially the powercap sysfs file interface

yes and no: powercap sysfs interface abstracts away MSR access, but otherwise it is almost 1-to-1 mapping of RAPL in hardware. This comes with multiple challenges for application developers:
- power measurement and power capping are lumped together despite having very different access profiles (as exemplified by having different permissions until recently) 
- application has to deal with RAPL zones which are not only confusing but I guess also irrelevant for most use cases (e.g. I might be interested in "overall system power", "CPU/DRAM power", or "power of my process" - but unlikely in something as cryptic and system-dependent as "PKG0_uncore power")
- "all or nothing": I cannot ask for aggregated/downsampled/per-process measurements, which would eliminate security problems (at least in practice)

**Comment by connorimes** [comment]
I understand your points and share some of your frustrations.  I think there are several things to keep in mind about the current state of affairs.

First - yes, the powercap sysfs interface is essentially is a 1-to-1 mapping, which is really what the OS *should* do whenever possible.  Additional layers of abstraction means additional code running in kernel space, which increases the likelihood of security issues and maintenance costs.  Additionally, abstractions are imperfect, so some applications would want to get around the abstraction because the hardware would provide a feature that the abstraction hides or makes more difficult to use. So - also yes - this pushes challenges and complexity to userspace/applications, but that's how it should be.

Second - just because measurements for other components like core/uncore/DRAM are irrelevant for your (or even "most") use cases doesn't negate their value entirely.  The OS developers present a unified interface and leave it to software to find and use what it wants.  Similar to my first point, this is better than the alternative of having OS devs create separate interfaces for many different use cases, which is generally poor SW design, far less maintainable, and would almost certainly miss some use cases.  I'm not claiming that the powercap sysfs interface is the best design, but I think we can understand why it is what it is.

Third - getting/setting a power cap and reading/resetting an energy consumption counter can be done together or independently (not everybody cares about both), but they are related.  The *permissions* (or *privileges*) needed to read and/or write to these values is independent of their utility, and only matters for security concerns that exist because the CPUs (and other affected components) are shared resources, but this is an orthogonal concern to the filesystem structure.

Finally - w.r.t. aggregating/downsampling/etc to resolve security questions - how do we know what the right balances are?  I'm not sure (and doubt) there are provably safe choices except disabling access entirely.  E.g., downsampling might improve security to make existing attacks fail, but a sampling rate that seems safe today might be found tomorrow to be insecure against a better attack.  That seems to (at least in part) be how we got in today's situation to begin with.

I would address per-process measurements separately.  I believe there are userspace profiles that can provide estimates for this. Doing this more accurately in the kernel would likely require integration with the scheduler as the hardware itself is not process-aware.  Of course I would also like such a feature, but it would also be very difficult to do in a way that supports the range of processors that Linux supports, and would likely introduce unacceptable overhead for most users.  I believe there would also be security concerns with this that would need to be addressed.

Anyway, I hope that's helpful and not too much rambling.  Cheers.

**Comment by amkozlov** [comment]
Thank you for your detailed comments, they are indeed very helpful! 

I guess, the very purpose of OS is to provide abstraction of hardware. In practice, the level of this abstraction is a tradeoff between (expected) utility and maintenance/security overhead. I/O is a good example: as application programmer, I don't have to care about keyboard/disk/display HW specifics, [even if it comes with potential security risks](https://en.wikipedia.org/wiki/C_standard_library#Concepts,_problems_and_workarounds).

Also it's worth distinguishing OS kernel and OS API (=POSIX/standard library). The functionality I described above should (ideally) be part of the latter. More realistically, it could be a 3rd party module like `msr-safe`. But either way, it would still be some privileged code "leaking" data into userspace. 

> Finally - w.r.t. aggregating/downsampling/etc to resolve security questions - how do we know what the right balances are? I'm not sure (and doubt) there are provably safe choices except disabling access entirely. E.g., downsampling might improve security to make existing attacks fail, but a sampling rate that seems safe today might be found tomorrow to be insecure against a better attack. That seems to (at least in part) be how we got in today's situation to begin with.

There is no such thing as absolute safety :) 

Example: as unprivileged user, I can see CPU% and memusage both system-wide and per-process, but more detailed performance stats (CPU cycles, page-faults, cache misses etc.) are typically root-only.
 
> I would address per-process measurements separately. I believe there are userspace profiles that can provide estimates for this. 

If you have any details/links on that, please share!

> Doing this more accurately in the kernel would likely require integration with the scheduler as the hardware itself is not process-aware. Of course I would also like such a feature, but it would also be very difficult to do in a way that supports the range of processors that Linux supports, and would likely introduce unacceptable overhead for most users. I believe there would also be security concerns with this that would need to be addressed.

Perfectionism is the root of all evil. Take `intel_rapl`, it supports Intel CPUs only, is hard to use and allegedly had a serious security flaw until recently. I'm pretty sure one can do better than that...
